airflow/contrib/hooks/gcp_vision_hook.py

```diff
@@ -0,0 +1,245 @@
+
+        return product_set_id
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def get_product_set(self, location, product_set_id, project_id=None,
+                        retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetGetOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_SET_NAME_TEMPLATE.format(project_id, location, product_set_id)
+        self.log.info('Retrieving ProductSet: %s', name)
+        response = self._handle_request(lambda **kwargs: client.get_product_set(**kwargs), name=name,
+                                        retry=retry, timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet retrieved.')
+        self.log.debug('ProductSet retrieved:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def update_product_set(self, product_set, location=None, product_set_id=None, update_mask=None,
+                           project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetUpdateOperator`
+        """
+        client = self.get_client()
+        product_set = self._determine_name(product_set, product_set_id, location, project_id)
+        self.log.info('Updating ProductSet: %s', product_set.name)
+        response = self._handle_request(lambda **kwargs: client.update_product_set(**kwargs),
+                                        product_set=product_set, update_mask=update_mask, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet updated: %s', response.name if response else '')
+        self.log.debug('ProductSet updated:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def delete_product_set(self, location, product_set_id, project_id=None, retry=None, timeout=None,
+                           metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetDeleteOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_SET_NAME_TEMPLATE.format(project_id, location, product_set_id)
+        self.log.info('Deleting ProductSet: %s', name)
+        response = self._handle_request(lambda **kwargs: client.delete_product_set(**kwargs), name=name,
+                                        retry=retry, timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet with the name [%s] deleted.', name)
+        return response
+
+    @staticmethod
+    def _determine_name(entity, entity_id, location, project_id):
+        is_product = type(entity) == Product
```

Comment:
This is not the right solution. You should take logic to a separate class and parameterize it, because you are now creating a code that is very hard to develop and maintain.

Reply:
This function does not do one thing, but it does two at the same time, but the problem is just similar.

Reply:
The logic is too complicated in my opinion. I have a big problem to understand all the cases that are included in this code.  Look at this: https://gitlab.polidea.com/snippets/18

Reply:
I respect that you may have a different opinion and invite you to a discussion.

Reply:
Look at:  https://gitlab.polidea.com/snippets/19